File: core/src/main/java/y2k/joyreactor/services/ProfileService.java
package y2k.joyreactor.services;

import rx.Observable;
import y2k.joyreactor.Profile;
import y2k.joyreactor.common.ObservableUtils;
import y2k.joyreactor.http.HttpClient;
import y2k.joyreactor.services.requests.LoginRequestFactory;
import y2k.joyreactor.services.requests.ProfileRequestFactory;

/**
 * Created by y2k on 11/25/15.
 */
public class ProfileService {

    private ProfileRequestFactory profileRequestFactory;
    private LoginRequestFactory loginRequestFactory;

    public ProfileService(ProfileRequestFactory profileRequestFactory, LoginRequestFactory loginRequestFactory) {
        this.profileRequestFactory = profileRequestFactory;
        this.loginRequestFactory = loginRequestFactory;
    }

    public Observable<?> login(String username, String password) {
        return loginRequestFactory.request(username, password);
    }

    public Observable<Profile> get() {
        return profileRequestFactory.request();
    }

    public Observable<?> logout() {
        return ObservableUtils.create(() -> HttpClient.getInstance().clearCookies());
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/ProfileService.kt
package y2k.joyreactor.services

import y2k.joyreactor.common.async.*
import y2k.joyreactor.common.http.HttpClient
import y2k.joyreactor.model.Profile
import y2k.joyreactor.services.requests.LoginRequestFactory

/**
 * Created by y2k on 11/25/15.
 */
class ProfileService(
    private val httpClient: HttpClient,
    private val requestProfile: (String) -> CompletableFuture<Profile>,
    private val loginRequestFactory: LoginRequestFactory,
    private val requestMyName: () -> CompletableFuture<String>) {

    fun login(username: String, password: String): CompletableFuture<*> {
        return loginRequestFactory.request(username, password)
    }

    fun getProfile(): CompletableFuture<Profile> {
        return requestMyName()
            .thenAsync { requestProfile(it) }
    }

    fun logout(): CompletableFuture<*> {
        return runAsync { httpClient.clearCookies() }
    }

    fun isAuthorized(): CompletableFuture<Boolean> {
        return requestMyName().then { true }.onError { false }
    }
}